@akerouanton looks like there's some (potentially) legit failures;

```
=== Failed
=== FAIL: amd64.integration-cli TestDockerSwarmSuite/TestOverlayAttachable (1.22s)
    docker_cli_swarm_test.go:402: assertion failed: error is not nil: exit status 125: 8e83551cb95e7f06866b463de38a149694b8ede41d088691ee01a233c7d066ca
        /usr/local/cli-integration/docker: Error response from daemon: could not find a network matching network mode ovnet: network ovnet not found.
        
    --- FAIL: TestDockerSwarmSuite/TestOverlayAttachable (1.22s)

=== FAIL: amd64.integration-cli TestDockerSwarmSuite/TestOverlayAttachableOnSwarmLeave (1.75s)
    docker_cli_swarm_test.go:421: assertion failed: error is not nil: exit status 125: 3ff1c5cde4b9564035c0399270b8a4fb67611378aa29b47832c827971ad50527
        /usr/local/cli-integration/docker: Error response from daemon: could not find a network matching network mode attovl: network attovl not found.
        
    --- FAIL: TestDockerSwarmSuite/TestOverlayAttachableOnSwarmLeave (1.75s)

=== FAIL: amd64.integration-cli TestDockerSwarmSuite/TestOverlayAttachableReleaseResourcesOnFailure (1.73s)
    docker_cli_swarm_test.go:446: assertion failed: error is not nil: exit status 125: 160bd7ef1111009c0dba21883f66ef6f698478d75b5403de5b6b384b4d71781f
        /usr/local/cli-integration/docker: Error response from daemon: could not find a network matching network mode ovnet: network ovnet not found.
        
    --- FAIL: TestDockerSwarmSuite/TestOverlayAttachableReleaseResourcesOnFailure (1.73s)

=== FAIL: amd64.integration-cli TestDockerSwarmSuite/TestSwarmContainerAttachByNetworkId (1.78s)
    docker_cli_swarm_test.go:370: assertion failed: error is not nil: exit status 125: be84a50271720d08a141fcee7127feaeca79cf28a131e8bb8e673e60a53f78d3
        /usr/local/cli-integration/docker: Error response from daemon: could not find a network matching network mode 6sux43f5ze7wjk74fcniahx9j: network 6sux43f5ze7wjk74fcniahx9j not found.
        
    --- FAIL: TestDockerSwarmSuite/TestSwarmContainerAttachByNetworkId (1.78s)

=== FAIL: amd64.integration-cli TestDockerSwarmSuite/TestSwarmContainerAutoStart (1.75s)
    docker_cli_swarm_test.go:322: assertion failed: error is not nil: exit status 125: 9d2314700c35d5bdb113da738d03ca27505bbf6b212a0d94a258bc7a861ac919
        /usr/local/cli-integration/docker: Error response from daemon: could not find a network matching network mode foo: network foo not found.
        
    --- FAIL: TestDockerSwarmSuite/TestSwarmContainerAutoStart (1.75s)

=== FAIL: amd64.integration-cli TestDockerSwarmSuite/TestSwarmContainerEndpointOptions (4.33s)
    docker_cli_swarm_test.go:344: assertion failed: error is not nil: exit status 125: 25cad7271c2543866de6f9cb6aff7c7a01c16a0b23081de29b4ee73869125c8c
        Unable to find image 'busybox:glibc' locally
        glibc: Pulling from library/busybox
        3f4d90098f5b: Pulling fs layer
        3f4d90098f5b: Verifying Checksum
        3f4d90098f5b: Download complete
        3f4d90098f5b: Pull complete
        Digest: sha256:7338d0c72c655d6534670ed72590cdc99afc9c559958c47eb3a5a17a7520a75c
        Status: Downloaded newer image for busybox:glibc
        /usr/local/cli-integration/docker: Error response from daemon: could not find a network matching network mode foo: network foo not found.
        
    --- FAIL: TestDockerSwarmSuite/TestSwarmContainerEndpointOptions (4.33s)

=== FAIL: amd64.integration-cli TestDockerSwarmSuite (1441.65s)
```

--- a/daemon/container_operations.go
+++ b/daemon/container_operations.go
@@ -691,6 +691,12 @@ func (daemon *Daemon) connectToNetwork(cfg *config.Config, container *container.
 	}
 	nwName := n.Name()
 
+	if idOrName != container.HostConfig.NetworkMode.NetworkName() {
+		if err := daemon.normalizeNetMode(container); err != nil {
+			return err
+		}
+	}
+
 	var operIPAM bool
 	if nwCfg != nil {
 		if epConfig, ok := nwCfg.EndpointsConfig[nwName]; ok {
@@ -884,6 +890,25 @@ func (daemon *Daemon) tryDetachContainerFromClusterNetwork(network *libnetwork.N
 	daemon.LogNetworkEventWithAttributes(network, events.ActionDisconnect, map[string]string{
 		"container": container.ID,
 	})
+}
+
+// normalizeNetMode checks whether the network mode references a network by a partial ID. In that case, it replaces the
+// partial ID with the full network ID.
+// TODO(aker): transform ID into name when the referenced network is one of the predefined.
+func (daemon *Daemon) normalizeNetMode(container *container.Container) error {
+	if container.HostConfig.NetworkMode.IsUserDefined() {
+		netMode := container.HostConfig.NetworkMode.NetworkName()
+		nw, err := daemon.FindNetwork(netMode)
+		if err != nil {
+			return fmt.Errorf("could not find a network matching network mode %s: %w", netMode, err)
+		}
+
+		if netMode != nw.ID() && netMode != nw.Name() {
+			container.HostConfig.NetworkMode = containertypes.NetworkMode(nw.ID())
+		}
+	}
+
+	return nil
 }
 
 func (daemon *Daemon) initializeNetworking(cfg *config.Config, container *container.Container) error {

--- a/daemon/network.go
+++ b/daemon/network.go
@@ -870,7 +870,8 @@ func buildCreateEndpointOptions(c *container.Container, n *libnetwork.Network, e
 	// to which container was connected to on docker run.
 	// Ideally all these network-specific endpoint configurations must be moved under
 	// container.NetworkSettings.Networks[n.Name()]
-	if nwName == c.HostConfig.NetworkMode.NetworkName() || (nwName == defaultNetName && c.HostConfig.NetworkMode.IsDefault()) {
+	netMode := c.HostConfig.NetworkMode
+	if nwName == netMode.NetworkName() || n.ID() == netMode.NetworkName() || (nwName == defaultNetName && netMode.IsDefault()) {
 		if c.Config.MacAddress != "" {
 			mac, err := net.ParseMAC(c.Config.MacAddress)
 			if err != nil {

--- a/integration/container/run_linux_test.go
+++ b/integration/container/run_linux_test.go
@@ -286,3 +286,29 @@ func TestRunWithAlternativeContainerdShim(t *testing.T) {
 
 	assert.Equal(t, strings.TrimSpace(b.String()), "Hello, world!")
 }
+
+func TestMacAddressIsAppliedToMainNetworkWithShortID(t *testing.T) {
+	skip.If(t, testEnv.IsRemoteDaemon)
+	skip.If(t, testEnv.DaemonInfo.OSType != "linux")
+
+	ctx := testutil.StartSpan(baseContext, t)
+
+	d := daemon.New(t)
+	d.StartWithBusybox(ctx, t)
+	defer d.Stop(t)
+
+	apiClient := d.NewClientT(t)
+
+	n := net.CreateNoError(ctx, t, apiClient, "testnet", net.WithIPAM("192.168.101.0/24", "192.168.101.1"))
+
+	cid := container.Run(ctx, t, apiClient,
+		container.WithImage("busybox:latest"),
+		container.WithCmd("/bin/sleep", "infinity"),
+		container.WithStopSignal("SIGKILL"),
+		container.WithNetworkMode(n[:10]),
+		container.WithMacAddress("02:42:08:26:a9:55"))
+	defer container.Remove(ctx, t, apiClient, cid, types.ContainerRemoveOptions{Force: true})
+
+	c := container.Inspect(ctx, t, apiClient, cid)
+	assert.Equal(t, c.NetworkSettings.Networks["testnet"].MacAddress, "02:42:08:26:a9:55")
+}

--- a/integration/internal/container/container.go
+++ b/integration/internal/container/container.go
@@ -154,3 +154,19 @@ func demultiplexStreams(ctx context.Context, resp types.HijackedResponse) (strea
 	wg.Wait()
 	return s, err
 }
+
+func Remove(ctx context.Context, t *testing.T, apiClient client.APIClient, container string, options types.ContainerRemoveOptions) {
+	t.Helper()
+
+	err := apiClient.ContainerRemove(ctx, container, options)
+	assert.NilError(t, err)
+}
+
+func Inspect(ctx context.Context, t *testing.T, apiClient client.APIClient, containerRef string) types.ContainerJSON {
+	t.Helper()
+
+	c, err := apiClient.ContainerInspect(ctx, containerRef)
+	assert.NilError(t, err)
+
+	return c
+}

--- a/integration/internal/container/ops.go
+++ b/integration/internal/container/ops.go
@@ -280,3 +280,15 @@ func WithPIDMode(mode container.PidMode) func(c *TestContainerConfig) {
 		c.HostConfig.PidMode = mode
 	}
 }
+
+func WithStopSignal(stopSignal string) func(c *TestContainerConfig) {
+	return func(c *TestContainerConfig) {
+		c.Config.StopSignal = stopSignal
+	}
+}
+
+func WithMacAddress(address string) func(c *TestContainerConfig) {
+	return func(c *TestContainerConfig) {
+		c.Config.MacAddress = address
+	}
+}

